lib/tree.rb

```diff
@@ -16,42 +16,132 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # Space Complexity: [O(1)-if it's a loop] - why do I wanna say O(n)?? help me. 
+  # Try with a while loop
+  def add(key, value = nil)
```

Comment:
👍 The time complexity is O(n) if the tree is balanced and O(log n) if it is balanced.
Since you're doing recursion in `add_helper` the space complexity is the same as the time.